### PR TITLE
Support oracle connect descriptor url

### LINKED ISSUE #7195: Support Oracle RAC
When the url type like this

 _jdbc:oracle:thin:@(DESCRIPTION =(ADDRESS = (PROTOCOL = TCP)(HOST = 172.16.0.12)(PORT = 1521))(ADDRESS = (PROTOCOL = TCP)(HOST = 172.16.0.22)(PORT = 1521))(LOAD_BALANCE = yes)(FAILOVER = ON)(CONNECT_DATA =(SERVER = DEDICATED)(SERVICE_NAME = rac)(FAILOVER_MODE=(TYPE = SELECT)(METHOD = BASIC)(RETIRES = 20)(DELAY = 15))))

it throws UnrecognizedDatabaseURLException.

### PR DESCRIPTION
**Pull request by xbkaishui**
Fixes #7195 

Changes proposed in this pull request:
- support oracle rac (connect descriptor url style) see [detail](https://docs.oracle.com/en/database/oracle/oracle-database/18/netag/identifying-and-accessing-database.html#GUID-2BDF9E52-4147-4F46-84E2-A5AE1018A373)

### CODE REVIEW

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/OracleDataSourceMetaData.java [hunk 1]
```diff
@@ -33,24 +33,39 @@
...
-    private final String catalog;
+    private String catalog;
     
-    private final String schema;
+    private String schema;
     
-    private final Pattern pattern = Pattern.compile("jdbc:oracle:(thin|oci|kprb):@(//)?([\\w\\-\\.]+):?([0-9]*)[:/]([\\w\\-]+)", Pattern.CASE_INSENSITIVE);
+    private final Pattern thinUrlPattern = Pattern.compile("jdbc:oracle:(thin|oci|kprb):@(//)?([\\w\\-\\.]+):?([0-9]*)[:/]([\\w\\-]+)", Pattern.CASE_INSENSITIVE);
+    
+    private final Pattern connectDescriptorUrlPattern = Pattern.compile("jdbc:oracle:(thin|oci|kprb):@[(\\w\\s=)]+HOST\\s*=\\s*(\\d+(\\."
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/OracleDataSourceMetaData.java:46]: Could we merge two of the patterns?

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/OracleDataSourceMetaData.java [hunk 2]
```diff
@@ -33,24 +33,39 @@
...
-    private final String hostName;
+    private String hostName;
     
-    private final int port;
+    private int port;
     
-    private final String catalog;
+    private String catalog;
     
-    private final String schema;
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/OracleDataSourceMetaData.java:42]: Why did you remove the `final`?

**Review by tristaZero — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/OracleDataSourceMetaData.java (hunk 1 above)
Comment (xbkaishui) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/OracleDataSourceMetaData.java:46]: hard be, this is refer to oracle JDBC driver's implementation

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/OracleDataSourceMetaData.java (hunk 2 above)
Comment (xbkaishui) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/OracleDataSourceMetaData.java:42]: it is used in method matchWithConnectDescriptorPattern, which is not in constructor method

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/OracleDataSourceMetaData.java (hunk 1 above)
Comment (xbkaishui) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/OracleDataSourceMetaData.java:46]: As you know. those are totally different, those have different patterns and different group counts

Comment (xbkaishui) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/OracleDataSourceMetaData.java:46]: done

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/OracleDataSourceMetaData.java (hunk 2 above)
Comment (xbkaishui) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/OracleDataSourceMetaData.java:42]: done

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/OracleDataSourceMetaData.java [hunk 3]
```diff
@@ -41,12 +41,23 @@
     
     private final String schema;
     
-    private final Pattern pattern = Pattern.compile("jdbc:oracle:(thin|oci|kprb):@(//)?([\\w\\-\\.]+):?([0-9]*)[:/]([\\w\\-]+)", Pattern.CASE_INSENSITIVE);
+    private final Pattern thinUrlPattern = Pattern.compile("jdbc:oracle:(thin|oci|kprb):@(//)?([\\w\\-\\.]+):?([0-9]*)[:/]([\\w\\-]+)", Pattern.CASE_INSENSITIVE);
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/OracleDataSourceMetaData.java:44]: Hi, why did you rename it as `thinUrlPattern`? Did you quote it somewhere?

**Review by tristaZero — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/OracleDataSourceMetaData.java [hunk 4]
```diff
@@ -41,12 +41,23 @@
...
+            + "((2(5[0-5]|[0-4]\\d))|[0-1]?\\d{1,2})){3}).*PORT\\s*=\\s*(\\d+).*SERVICE_NAME\\s*=\\s*(\\w+)\\)");
     
     public OracleDataSourceMetaData(final String url, final String username) {
-        Matcher matcher = pattern.matcher(url);
+        Matcher matcher = thinUrlPattern.matcher(url);
         if (!matcher.find()) {
-            throw new UnrecognizedDatabaseURLException(url, pattern.pattern());
+            matcher = connectDescriptorUrlPattern.matcher(url);
+            if (!matcher.find()) {
+                throw new UnrecognizedDatabaseURLException(url, connectDescriptorUrlPattern.pattern());
```
Comment (tristaZero) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/database/metadata/dialect/OracleDataSourceMetaData.java:54]: It seems not so concise. How about this programing clip?
```java
Collection<Matcher> matchers;
for each `matchers`
if each.find() {
}
end
throw new Exception
```

**Review by tristaZero — CHANGES_REQUESTED** [review]

**Review by tristaZero — APPROVED** [review]

**Comment by tristaZero** [comment]
Very appreciated your continued work. @xbkaishui Your contribution is exactly an important part of this community!
Have a lovely weekend. 😉
